lib/puppet/http/service/compiler.rb

```diff
@@ -129,6 +129,70 @@ def post_catalog(name, facts:, environment:, configured_environment: nil, transa
     [response, deserialize(response, Puppet::Resource::Catalog)]
   end
 
+  # TODO; document this API once it is finalized
+  def post_catalog4(certname, persistence, environment, facts: nil, trusted_facts: nil, transaction_uuid: nil, job_id: nil, options: nil)
+    body = {
+      certname: certname,
+      persistence: persistence,
+      environment: environment,
+      transaction_uuid: transaction_uuid,
+      job_id: job_id,
+      options: options
+    }
+    body.merge!({ facts: { values: facts }}) unless facts.nil?
+    body.merge!({ trusted_facts: { values: trusted_facts }}) unless trusted_facts.nil?
+    headers = {
+      'accept-encoding' => Puppet::Network::HTTP::Compression::ACCEPT_ENCODING,
+      'Content-Type' => 'text/json',
+      'Accept' => 'application/json',
+      Puppet::Network::HTTP::HEADER_PUPPET_VERSION => Puppet.version
+    }
+
+    # CODEREVIEW: The base url has v3 baked in and does not appear to be easily configured here
+    # Is this somewhat hacky work around acceptable?
+    v3_regex = /\/v3$/
+    url = with_base_url("")
+    if url.path =~ v3_regex
+      url.path = url.path.gsub(v3_regex, '') + '/v4/catalog'
+    else
+      #TODO: raise error
+    end
+
+    response = @client.post(
+      url,
+      body.to_json,
+      headers: headers
+    )
+
+    # CODEREVIEW: The following code was lifted from (a deprecated) method in
+    # https://github.com/puppetlabs/puppet/blob/bc4f0a44c60069d42f7d5ddb02b4c026fcaf522a/lib/puppet/network/http/compression.rb#L20-L34
+    # I assume we want to compress the catalog response over the network for performance reasons. What is the preferred way to handle 
+    # decompressing without the deprecated method?
+    require 'zlib'
+    require 'stringio'
+
+    uncompressed = case response['content-encoding']
+      when 'gzip'
+        # ZLib::GzipReader has an associated encoding, by default Encoding.default_external
+        Zlib::GzipReader.new(StringIO.new(response.body), :encoding => Encoding::BINARY).read
+      when 'deflate'
+        Zlib::Inflate.new.inflate(response.body)
+      when nil, 'identity'
+        response.body
+      else
+        raise Net::HTTPError.new(_("Unknown content encoding - %{encoding}") % { encoding: response['content-encoding'] }, response)
+      end
```

Comment:
The http client will automatically decompress the response body, so you should be able to remove this.